docs/GenCodec.md

```diff
@@ -64,13 +64,51 @@ bound to any format. It only depends on the fact that this format is capable of
 * `Char`s, `String`s, `Boolean`s and `null`s
```

Comment:
Now that we have both `.nextField` and `.peekField`, we need an example more complex than a single field class and less automatically handled than a case class. I've seen a lot of people struggle with creating 2-field class codec and it's just getting harder.

More generally, should I `peek` or should I `next`? It appears to me that GenCodec won't exactly be completely Input/Output indepedent.

Reply:
If you want to write a custom codec that creates an object with two-three fields then the way to go is to implement `apply`/`unapply` in companion object or use [fromApplyUnapplyProvider](http://avsystem.github.io/scala-commons/api/com/avsystem/commons/serialization/GenCodec$.html#fromApplyUnapplyProvider[T](applyUnapplyProvider:Any):com.avsystem.commons.serialization.GenCodec[T]). This is now very clearly recommended in the docs.

You should almost never need to implement an object codec manually. If you *really* need to do it then the backend-independent way of doing it would be:

```scala
class Klass(val int: Int, val str: String)
object Klass {
  private def getField(input: ObjectInput, name: String):
    input.peekField(name).orElse(input.nextField().opt).filter(_.fieldName == name)
      .getOrElse(throw new ReadFailure(s"expected field $name")

  implicit val codec: GenCodec[Klass] = GenCodec.createNullableObject(
    input => {
      val int = getField(input, "int").readInt()
      val str = getField(input, "str").readString()
      new Klass(int, str)
    },
    (klass, output) => {
      output.writeField("int").writeInt(klass.int)
      output.writeField("str").writeString(klass.str)
    }
  )
}
```

but really, this should be an absolute last resort.

Reply:
`Input` and `Output` interfaces are generally optimized for ease of *implementation* and performance rather than ease of use.